New board: AIRLink board configuration added

aviaks opened the pull request:
Added AIRLink board configuration.

dagar commented on boards/sky-drones/smartap-airlink/init/rc.board_sensors:
```diff
@@ -0,0 +1,34 @@
+#bmp388 -I start
+
+# External HMC5983
+hmc5883 -T -X -R 4 start
```
Is this a specific sensor that is normally present and expected to have that rotation?

aviaks replied:
This is a magnetometer on externally connected GPS module which is normally used. The rotation is related to the sensor on that GPS module.

dagar commented:
At the moment we have some CI duplication between Jenkins and Github actions, can you also add the board here?

https://github.com/PX4/PX4-Autopilot/blob/680fe86c0535ede9fe7bcb0cebfdf6c94a355e8e/.github/workflows/compile_nuttx.yml#L63-L65

dagar commented:
Overall the change looks good. Do you have any schematic or basic pin out that's available? 

If you have any flight log (even from bench testing) please share and I'll verify that everything is running as expected with no errors.

dagar commented:
Minor whitespace error failure in CI. https://github.com/PX4/PX4-Autopilot/runs/6141267293?check_suite_focus=true
![Screenshot from 2022-04-23 13-05-40](https://user-images.githubusercontent.com/84712/164916243-eb90d1ca-fa43-46a1-9aee-0f02dec7c509.png)

aviaks commented:
> At the moment we have some CI duplication between Jenkins and Github actions, can you also add the board here?

Sorry for forgetting to add this, now added.

> Minor whitespace error failure in CI. https://github.com/PX4/PX4-Autopilot/runs/6141267293?check_suite_focus=true !

Thanks! Fixed now. 

> Overall the change looks good. Do you have any schematic or basic pin out that's available?
> 
> If you have any flight log (even from bench testing) please share and I'll verify that everything is running as expected with no errors.

Yes, I'll send the pinout on your email shortly. Please see the bench testing flight log attached - [log_3_2022-4-23-18-24-20.ulg.zip](https://github.com/PX4/PX4-Autopilot/files/8547697/log_3_2022-4-23-18-24-20.ulg.zip)

dagar commented on boards/sky-drones/smartap-airlink/init/rc.board_mavlink:
```diff
@@ -0,0 +1,7 @@
+#------------------------------------------------------------------------------
+
+# Start MAVLink on the USB port
+mavlink start -d /dev/ttyACM0
```
This can be skipped if you have VBUS. We monitor USB and only start the CDCACM driver once it's connected, and then based on actual serial traffic automatically start mavlink USB when we see a HEARTBEAT. Alternatively you can also launch NSH with 3 newlines, or use it as bridge between the GPS and ucenter.

aviaks replied:
Thanks, makes sense! Yes, we have VBUS. Now removed.

dagar commented on boards/sky-drones/smartap-airlink/nuttx-config/nsh/defconfig:
```diff
@@ -0,0 +1,294 @@
+CONFIG_STM32F7_SPI1_DMA_BUFFER=1024
+CONFIG_STM32F7_SPI2=y
+CONFIG_STM32F7_SPI2_DMA=y
+CONFIG_STM32F7_SPI2_DMA_BUFFER=4096
```
```suggestion
CONFIG_STM32F7_SPI2_DMA_BUFFER=1024
```

aviaks replied:
Thanks, done!

dagar commented:
Log looks ok, but you apparently have very little free memory and I'm not yet sure why. 

https://logs.px4.io/plot_app?log=b65ec398-780d-4b7e-89f9-9fb7c99d149f

Until we figure out you could lower the logger buffer a bit, maybe add `set LOGGER_BUF 32` in rc.board_defaults?

https://github.com/PX4/PX4-Autopilot/blob/680fe86c0535ede9fe7bcb0cebfdf6c94a355e8e/platforms/nuttx/init/stm32f7/rc.board_arch_defaults#L18

Your log has 4 mavlink instances running, what do you actually need/want by default? Run `mavlink status` to see.

dagar commented:
For reference on a Holybro v5x configured with the same airframe (4015) I'm seeing 110-120 kB of free RAM (~ 25%)

dagar commented:
I just noticed the eeprom driver failed to start.
![Screenshot from 2022-04-23 14-07-18](https://user-images.githubusercontent.com/84712/164934337-204e00e4-3d8b-40b3-901f-34c54efb9127.png)

Is this part physically present?

https://github.com/PX4/PX4-Autopilot/blob/b5386ca0665314a57ebbdd85897ed99088c713cb/boards/sky-drones/smartap-airlink/src/mtd.cpp#L45-L48

dagar commented on boards/sky-drones/smartap-airlink/init/rc.board_sensors:
```diff
@@ -0,0 +1,34 @@
+board_adc start
+
+# Internal SPI bus MPU-9250 / SPI 2 [FPC]
+mpu9250 -s -b 2 -M -R 6 start
```
```suggestion
mpu9250 -s -b 2 -R 6 start
```

I would skip the mpu9250 mag if you have another internal magnetometer working. Long story, but when the mpu9250 ak8963 is enabled is disrupts the FIFO output and you don't get clean 8 kHz raw data.

dagar replied:
If you plot `vehicle_imu_status.gyro_raw_rate_hz` or look at `sensors status` you'll be able to see the difference.

aviaks replied:
Thanks! Then it makes sense indeed. I would still prefer to have magnetometer enabled on one mpu9250 but not on both for redundancy in addition to onboard bmm150.

aviaks commented:
> Log looks ok, but you apparently have very little free memory and I'm not yet sure why.
> 
> https://logs.px4.io/plot_app?log=b65ec398-780d-4b7e-89f9-9fb7c99d149f

Yes, this is strange, trying to figure out why..

> Until we figure it out you could lower the logger buffer a bit, maybe add `set LOGGER_BUF 32` in rc.board_defaults?
> 
> https://github.com/PX4/PX4-Autopilot/blob/680fe86c0535ede9fe7bcb0cebfdf6c94a355e8e/platforms/nuttx/init/stm32f7/rc.board_arch_defaults#L18

Yes, done. 

> Your log has 4 mavlink instances running, what do you actually need/want by default? Run `mavlink status` to see.

Three if them are - USB, Telemetry 1 external UART port and UART to mission computer. The fourth looks like duplicated USB probably due to explicitly started in rc.board_mavlink.. 

<img width="404" alt="image" src="https://user-images.githubusercontent.com/1250143/164934506-e630a650-4c56-4429-a3a4-ca9e3013d680.png">

dagar commented:
How does it look now? Holybro px4_fmu-v5x for comparison.

``` Console

nsh> free
                   total       used       free    largest  nused  nfree
        Umem:     431488     320272     111216     110144   1387     68
        Prog:    2097152    2097152          0          0     13      0

nsh> listener cpuload

TOPIC: cpuload
 cpuload
    timestamp: 1075749214 (0.470215 seconds ago)
    load: 0.6990
    ram_usage: 0.7408
```

aviaks commented:
> How does it look now? Holybro px4_fmu-v5x for comparison.
> 
> ```
> nsh> free
>                    total       used       free    largest  nused  nfree
>         Umem:     431488     320272     111216     110144   1387     68
>         Prog:    2097152    2097152          0          0     13      0
> 
> nsh> listener cpuload
> 
> TOPIC: cpuload
>  cpuload
>     timestamp: 1075749214 (0.470215 seconds ago)
>     load: 0.6990
>     ram_usage: 0.7408
> ```

This is with logger buffer set to 32:

<img width="565" alt="image" src="https://user-images.githubusercontent.com/1250143/164934665-05474f97-e254-4e48-977d-027199d8a809.png">

aviaks commented:
> Three if them are - USB, Telemetry 1 external UART port and UART to mission computer. The fourth looks like duplicated USB probably due to explicitly started in rc.board_mavlink..
> 
> <img alt="image" width="404" src="https://user-images.githubusercontent.com/1250143/164934506-e630a650-4c56-4429-a3a4-ca9e3013d680.png">

Removing mavlink start from rc.board_mavlink removed the unused instance, now there are 3 of them in mavlink status (USB, Telemetry 1, telemetry to mission computer). So it looks good now.

dagar commented:
Ok, that looks tolerable now, that mavlink USB instance is quite heavy. You shouldn't have been able to start duplicate instances on the same device, so I'll see why that was able to slip through.

Multi-EKF is also quite expensive, if you need more resources for whatever reason in the future an easy change will be EKF2_MULTI_IMU 3->2 or disabling it entirely (SENS_IMU_MODE).

aviaks commented:
> I just noticed the eeprom driver failed to start. ![Screenshot from 2022-04-23 14-07-18](https://user-images.githubusercontent.com/84712/164934337-204e00e4-3d8b-40b3-901f-34c54efb9127.png)
> 
> Is this part physically present?
> 
> https://github.com/PX4/PX4-Autopilot/blob/b5386ca0665314a57ebbdd85897ed99088c713cb/boards/sky-drones/smartap-airlink/src/mtd.cpp#L45-L48

Thanks for pointing to this! There is only one EEPROM (base), the second one was initialised in the code, but not present. Now fixed.

aviaks commented:
> Ok, that looks tolerable now, that mavlink USB instance is quite heavy. You shouldn't have been able to start duplicate instances on the same device, so I'll see why that was able to slip through.

Thanks! Yes, that worked out for some reason but it showed zero rate data going through.. 

> Multi-EKF is also quite expensive, if you need more resources for whatever reason in the future an easy change will be EKF2_MULTI_IMU 3->2 or disabling it entirely (SENS_IMU_MODE).

Good, thanks! Will keep this in mind!

aviaks commented:
Unfortunately, the build is failing here: 
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/1250143/164941251-1430105a-5417-477f-8cce-2b6e6688b04e.png">
However, builds fine on my laptop. @dagar could you suggest what's wrong please?

dagar commented:
> However, builds fine on my laptop. @dagar could you suggest what's wrong please?

The build artifacts are probably caught in some bad intermediate state from previous partial builds. For now I would just clean (`make distclean`) and try again. If it's repeatable we can dig into it and find the root cause.

aviaks commented:
> > However, builds fine on my laptop. @dagar could you suggest what's wrong please?
> 
> The build artifacts are probably caught in some bad intermediate state from previous partial builds. For now I would just clean (`make distclean`) and try again. If it's repeatable we can dig into it and find the root cause.
> 
> **EDIT:** It's failing on GitHub actions, I'll take a look.

Thanks Daniel!

dagar commented on boards/sky-drones/smartap-airlink/nuttx-config/nsh/defconfig:
```diff
@@ -0,0 +1,294 @@
+# CONFIG_NSH_DISABLE_TIME is not set
+CONFIG_ARCH="arm"
+CONFIG_ARCH_BOARD_CUSTOM=y
+CONFIG_ARCH_BOARD_CUSTOM_DIR="../../../../boards/px4/fmu-v5x/nuttx-config"
```
```suggestion
CONFIG_ARCH_BOARD_CUSTOM_DIR="../../../../boards/sky-drones/smartap-airlink/nuttx-config"
```

dagar commented:
@aviaks this was the problem https://github.com/PX4/PX4-Autopilot/pull/19529/commits/5c58e615015c2fa90b2795f938c129615ccb87bd.

This should be good to go now. I'll squash and merge all commits into master once the build goes through cleanly.

aviaks commented:
> @aviaks this was the problem [5c58e61](https://github.com/PX4/PX4-Autopilot/commit/5c58e615015c2fa90b2795f938c129615ccb87bd).
> 
> This should be good to go now. I'll squash and merge all commits into master once the build goes through cleanly.

Ah, my fault, this was an easy one.. Many thanks @dagar for resolving this!
